### ISSUE TITLE
Search Pagination

### ISSUE THREAD

**Issue by eselimsen** [FlixCoder/fhir-sdk#34]
Hi @FlixCoder
Thanks for this useful crate. I've been using it for a web backend project and stumbled upon an issue, I would appreciate your input.
I'm trying to consume `Paged` resources manually, such as `search()` results to implement paging for infinite scrolling. However, the `Stream` returned from `search()` doesn't provide any breakpoint we can pause and resume later as needed. Also, as `Paged` and `read_generic()` are private I can't workaround by crating manual URL.

Do you have any suggestions to handle this?
:slightly_smiling_face:

**Comment by FlixCoder** [comment]
Indeed, there is no way to do manual paging right now. ^^ Paging is already part of the returned `Stream`. In order to implement infinite scrolling, you can easily utilize the stream though.

- One option is to just keep the stream alive and whenever you need more resources to show, just fetch the next 10 or 20 elements from the stream via `buffer.push(stream.try_next().await?)` ^^
- Another option is [`stream.try_chunks(10)`](https://docs.rs/futures/latest/futures/prelude/stream/trait.TryStreamExt.html#method.try_chunks). This just gives you an array of 10 items every time you get the next stream item.

**Comment by eselimsen** [comment]
Thank you for prompt reply. Would you consider making functions such as `read_generic` public, so that crate users can implement missing features manually? I want to use this crate on my project, and having access to lower bits would be a huge relief, allowing any custom or not-yet-implemented features. :slightly_smiling_face: 

I also would like to ask your thoughts about a PR like so:
```rust
pub struct SearchResult {
    pub result: Result<NamedResource, Error>,
    pub next_url: Option<Url>,
}


// Just to demo the idea; add another function returning a struct with results of "this page", 
//  and url to the next one, essentially the data content of `Paged`.
// Next url could be later used with `read_generic`, or this function can take an `Option<Url>`?.
impl Client<FhirR5> {
    pub search_raw(parameters) -> SearchResult { 
    }
}
```

**Comment by FlixCoder** [comment]
Hmm read_generic is not really suitable to be public, as it cannot be used for any request. I understand there should be a way to add custom logic using as much of this clients functionality as possible. The read_generic assumes 404 is not an error though, which might not be suitable.
Technically, any access it already possible by just using the inner reqwest client with your URL and deserializing the respone to the requested FHIR resource, but you would loose automatic authentication and retrying.

I think I would rather like to offer a function on the client to make any HTTP request, but using the client's HTTP request machinery. That would force users to manually handle the reqwest::Response, but they can also do custom POST requests for any custom FHIR operations they like.

Regarding custom search, why is it that you don't like .search().try_chunks() if I may ask?

Your SearchResult only contains one resource right? That is not what the FHIR server responds with. The FHIR server responds with a Bundle of resources, where the next url can simply be extracted [like so](https://github.com/FlixCoder/fhir-sdk/blob/c1c14656e6eaeb4ec7a82af875448d9e173b65f6/crates/fhir-sdk/src/client/r5/paging.rs#L154). So Bundle is kind of a SearchResult already, however it is of course more annoying to manually extract next URL and resources. Maybe an extension trait for Bundles would make sense for that?

Making too many inner implementation details public can harm future compatibility. Even though I don't really make any guarantees yet, I feel like there are better ways to offer the same freedom and functionality.

I could imagine adding something like this:
```rust
let response = fhir.send_custom_request(|http| {
    http
        .post("https://custom.search/?param=value")
        .header(ACCEPT, MIME_TYPE)
    }).await?;
let bundle: Bundle = response.json()?;
let next_url = bundle.find_next_url();
```
What would you thin about this?

Although of course, the Bundle might have custom resource links that are not resolved then and the error needs to be detected and handles based on the response status.

**Comment by eselimsen** [comment]
> I think I would rather like to offer a function on the client to make any HTTP request

Sounds good to me, my aim is to have the possibility to use idioms like new-type to fill missing features, and reduce duplication on consumer-side, benefitting from "upstream" (crate) fix and features as much as possible. For example, I see value in having the checks in `fetch_resource` (maybe make part of `Client`?), helpers like `find_next_page_url` or filtering mechanism implemented in `Paged` creation. 

Thinking of a low-level API, I'd think they're internal to the crate and the author don't make any guarantees, but you can still reach out there and do whatever you need when you must, it is technically possible, taking the risk of it being changed between versions. Maybe they would sit in a different module, or if it means too much restructuring, low-level bits could have a note indicating that they're subject to change and consumers are discouraged to use them? But still, you can use new-type or other means to get your needs done.

> Regarding custom search, why is it that you don't like .search().try_chunks() if I may ask?

Unless I'm missing something, it still means I need to keep Stream alive, right? I want to avoid it -as a HTTP server-, as much as possible. Also, that'd mean having containers for all resource types, correct? 
Like `person_streams`, `patient_streams`...
:thinking:

**Comment by FlixCoder** [comment]
I pushed new changes to main with `send_custom_request` and `BundleExt::next_page_url`. You can try it out with a git dependency on current main (maybe set a custom revision/hash, so new changes won't bother :D).

Once I have time, I will try to add more functionality like the bundle entry filtering by matches and more. This even helps my efforts to reduce code duplication, so that is nice. Want to work on that, too. But this all needs a lot of time, so gonna take a while until I have time and fun doing so. Once #19 is tackled, it is also way easier to add new things.

**Comment by eselimsen** [comment]
That looks useful, thanks!

Meanwhile, could you also checkout https://github.com/FlixCoder/fhir-sdk/pull/35? It proposes adding a new manually-paged search functionality. I tested with my infinite scroll use case, it works perfectly and exposes more info about pagination (such as `previous_url` and `total` when supported).

**Comment by FlixCoder** [comment]
So #41 now allows the following:

```rust
let page = client.search::<Patient>(SearchParameters::empty()).await?;
let next_page_url = page.next_page_url().unwrap().clone();
let total = page.total().unwrap();

let next_page = client.search_custom::<Patient>(|http| http.get(next_page_url)).await?;
// or
let next_page = page.next_page().await?;

// Get current matches, ignoring missing or incorrect resources.
let current_entries = next_page.matches().collect::<Vec<_>>();
// Get matches in the current response, resolving full_url.
let current_entries = next_page.matches_owned().try_collect::<Vec<_>>().await?;

// And still automatic paging possible.
let all_entries = page.all_matches().try_collect::<Vec<_>>().await?;
```
